go/fixchain/fix_error_test.go

```diff
@@ -37,9 +48,120 @@ func TestTypeString(t *testing.T) {
 		},
 	}
 
-	for _, test := range fixErrorTests {
+	for i, test := range typeStringTests {
 		if got, want := test.ferr.TypeString(), test.expected; got != want {
-			t.Errorf("TypeString() returned %s, expected %s.", got, want)
+			t.Errorf("#%d: TypeString() returned %s, expected %s.", i, got, want)
+		}
+	}
+}
+
+func TestString(t *testing.T) {
+	stringTests := []struct {
+		ferr *FixError
+		str  string
+	}{
+		{
+			&FixError{Type: None},
+			"None\n",
+		},
+		{
+			&FixError{
+				Type: LogPostFailed,
+				Cert: GetTestCertificateFromPEM(t, googleLeaf),
+				Chain: []*x509.Certificate{
+					GetTestCertificateFromPEM(t, googleLeaf),
+					GetTestCertificateFromPEM(t, thawteIntermediate),
+					GetTestCertificateFromPEM(t, verisignRoot),
+				},
+				URL:   "https://www.test.com",
+				Error: errors.New("Log Post Failed"),
+				Code:  404,
+			},
+			"LogPostFailed\n" +
+				"Status Code: 404\n" +
+				"Error: Log Post Failed\n" +
+				"URL: https://www.test.com\n" +
+				"Cert: " + googleLeaf +
+				"Chain: " + googleLeaf + thawteIntermediate + verisignRoot,
+		},
+	}
+
+	for i, test := range stringTests {
+		if got, want := test.ferr.String(), test.str; got != want {
+			t.Errorf("#%d: String() returned %s, expected %s.", i, got, want)
+		}
+	}
+}
+
+func TestMarshalJSON(t *testing.T) {
+	var encode = base64.StdEncoding.EncodeToString
+
+	marshalJSONTests := []struct {
+		ferr *FixError
+		json string
+	}{
+		{
+			&FixError{Type: None},
+			"{\"Type\":\"None\",\"Cert\":null,\"Chain\":null,\"URL\":\"\",\"Bad\":null,\"Error\":\"\",\"Code\":0}",
+		},
+		{
+			&FixError{
+				Type: LogPostFailed,
+				Cert: GetTestCertificateFromPEM(t, googleLeaf),
+				Chain: []*x509.Certificate{
+					GetTestCertificateFromPEM(t, googleLeaf),
+					GetTestCertificateFromPEM(t, thawteIntermediate),
+					GetTestCertificateFromPEM(t, verisignRoot),
+				},
+				URL:   "https://www.test.com",
+				Error: errors.New("Log Post Failed"),
+				Code:  404,
+			},
+			"{\"Type\":\"LogPostFailed\"," +
```

Comment:
Changed to marshal then unmarshal and check result matches the original

Note:  This involved adding Equal() and Unmarshal() functions in fix_error.go that need reviewing.